api/controller.py

```diff
@@ -1416,74 +1419,104 @@ def borrow(self, identifier_type, identifier, mechanism_id=None):
                 _("I've never heard of this work.")
             )
         if isinstance(result, ProblemDetail):
+            # There was a problem determining the appropriate
+            # LicensePool to use.
             return result
-        pool, mechanism = result
 
-        header = self.authorization_header()
-        credential = self.manager.auth.get_credential_from_header(header)
-        problem_doc = None
+        if isinstance(result, Loan):
+            # We already have a Loan, so there's no need to go to the API.
+            loan_or_hold = result
+            is_new = False
+        else:
+            # We need to actually go out to the API
+            # and try to take out a loan.
+            pool, mechanism = result
+            loan_or_hold, is_new = self._borrow(patron, credential, pool, mechanism)
+
+        if isinstance(loan_or_hold, ProblemDetail):
+            return loan_or_hold
+
+        # At this point we have either a loan or a hold. If a loan, serve
+        # a feed that tells the patron how to fulfill the loan. If a hold,
+        # serve a feed that talks about the hold.
+        response_kwargs = {}
+        if is_new:
+            response_kwargs['status'] = 201
+        else:
+            response_kwargs['status'] = 200
+        return LibraryLoanAndHoldAnnotator.single_item_feed(
+            self.circulation, loan_or_hold, **response_kwargs
+        )
```

Comment:
This is the same code as before, but the diff makes it look new because it now comes before the code in the helper method, where it used to come afterwards.